File: pos-python/db/venta_dao.py
from decimal import Decimal
from db.connection import DB


def insertar_venta(turno_id, cajero_id, items, subtotal, iva, total,
                   metodo_pago_id, monto_recibido, cambio, cliente_id=None):
    """Inserts venta + detalle_ventas in one transaction. Returns venta_id."""
    conn = DB.conn()
    try:
        with conn.cursor() as cur:
            cur.execute(
                """INSERT INTO ventas
                       (turno_id, cajero_id, cliente_id, subtotal, iva, total,
                        metodo_pago_id, monto_recibido, cambio)
                   VALUES (%s,%s,%s,%s,%s,%s,%s,%s,%s) RETURNING id""",
                (turno_id, cajero_id, cliente_id, subtotal, iva, total,
                 metodo_pago_id, monto_recibido, cambio)
            )
            venta_id = cur.fetchone()[0]

            for item in items:
                cur.execute(
                    """INSERT INTO detalle_ventas
                           (venta_id, producto_id, cantidad, precio_unitario, subtotal)
                       VALUES (%s,%s,%s,%s,%s)""",
                    (venta_id, item["producto_id"], item["cantidad"],
                     item["precio"], item["subtotal"])
                )
                cur.execute(
                    """UPDATE inventario
                       SET stock_actual = stock_actual - %s
                       WHERE producto_id = %s""",
                    (item["cantidad"], item["producto_id"])
                )

            cur.execute(
                """UPDATE turnos
                   SET venta_total = venta_total + %s,
                       efectivo_ingresado = efectivo_ingresado + %s,
                       cobro_tarjeta = cobro_tarjeta + %s
                   WHERE id = %s""",
                (total,
                 total if metodo_pago_id == 1 else Decimal("0"),
                 total if metodo_pago_id != 1 else Decimal("0"),
                 turno_id)
            )
        conn.commit()
        return venta_id
    except Exception:
        conn.rollback()
        raise


def find_by_turno(turno_id: int):
    return DB.fetchall(
        """SELECT v.id, v.total, v.subtotal, v.iva, mp.nombre AS metodo,
                  v.creado_en
           FROM ventas v
           JOIN metodos_pago mp ON mp.id = v.metodo_pago_id
           WHERE v.turno_id = %s
           ORDER BY v.creado_en DESC""",
        (turno_id,)
    )


def find_detalle(venta_id: int):
    return DB.fetchall(
        """SELECT p.nombre, dv.cantidad, dv.precio_unitario, dv.subtotal
           FROM detalle_ventas dv
           JOIN productos p ON p.id = dv.producto_id
           WHERE dv.venta_id = %s""",
        (venta_id,)
    )


def get_metodos_pago():
    rows = DB.fetchall("SELECT id, nombre FROM metodos_pago ORDER BY id")
    if not rows:
        DB.execute("INSERT INTO metodos_pago (nombre) VALUES ('EFECTIVO') ON CONFLICT DO NOTHING")
        DB.execute("INSERT INTO metodos_pago (nombre) VALUES ('TARJETA') ON CONFLICT DO NOTHING")
        rows = DB.fetchall("SELECT id, nombre FROM metodos_pago ORDER BY id")
    return rows


def resumen_global():
    return DB.fetchone(
        """SELECT COUNT(*) AS total_ventas,
                  COALESCE(SUM(total), 0) AS monto_total
           FROM ventas WHERE estado='COMPLETADA'"""
    )

File: pos-maui/Modelo/Dao/VentaDao.cs
using Dapper;
using PosMaui.Modelo.Entidad;

namespace PosMaui.Modelo.Dao;

public static class VentaDao
{
    public static int AbrirTurno(int cajeroId, decimal fondoInicial)
    {
        using var conn = ConexionBD.Abrir();
        string codigo = $"T-{DateTime.Now:yyyyMMdd-HHmmss}";
        return conn.QueryFirst<int>("""
            INSERT INTO turnos (codigo, cajero_id, fondo_inicial, estado)
            VALUES (@Codigo, @CajeroId, @Fondo, 'ABIERTO')
            RETURNING id
            """, new { Codigo = codigo, CajeroId = cajeroId, Fondo = fondoInicial });
    }

    public static int GuardarVenta(int turnoId, int cajeroId,
                                    decimal subtotal, decimal iva, decimal total,
                                    int metodoPagoId, decimal montoRecibido, decimal cambio,
                                    List<CartItem> items)
    {
        using var conn = (Npgsql.NpgsqlConnection)ConexionBD.Abrir();
        using var tx = conn.BeginTransaction();
        try
        {
            int ventaId = conn.QueryFirst<int>("""
                INSERT INTO ventas (turno_id, cajero_id, subtotal, iva, total,
                                    metodo_pago_id, monto_recibido, cambio, estado)
                VALUES (@TurnoId, @CajeroId, @Subtotal, @Iva, @Total,
                        @MetodoPagoId, @MontoRecibido, @Cambio, 'COMPLETADA')
                RETURNING id
                """, new { TurnoId = turnoId, CajeroId = cajeroId,
                           Subtotal = subtotal, Iva = iva, Total = total,
                           MetodoPagoId = metodoPagoId, MontoRecibido = montoRecibido,
                           Cambio = cambio }, tx);

            foreach (var item in items)
            {
                conn.Execute("""
                    INSERT INTO detalle_ventas (venta_id, producto_id, cantidad, precio_unitario, subtotal)
                    VALUES (@VentaId, @ProductoId, @Cantidad, @Precio, @Subtotal)
                    """, new { VentaId = ventaId, ProductoId = item.DbProductoId,
                               Cantidad = (int)item.Cantidad, Precio = item.Precio,
                               Subtotal = item.Subtotal }, tx);

                conn.Execute("""
                    UPDATE inventario SET stock_actual = stock_actual - @Cant
                    WHERE producto_id = @PId
                    """, new { Cant = (int)item.Cantidad, PId = item.DbProductoId }, tx);
            }

            tx.Commit();
            return ventaId;
        }
        catch { tx.Rollback(); throw; }
    }

    public static decimal VentasGlobales()
    {
        using var conn = ConexionBD.Abrir();
        return conn.QueryFirst<decimal>(
            "SELECT COALESCE(SUM(total),0) FROM ventas WHERE estado='COMPLETADA'");
    }

    public static int ContarTurnosActivos()
    {
        using var conn = ConexionBD.Abrir();
        return conn.QueryFirst<int>("SELECT COUNT(*) FROM turnos WHERE estado='ABIERTO'");
    }

    public static int ContarCortes()
    {
        using var conn = ConexionBD.Abrir();
        return conn.QueryFirst<int>("SELECT COUNT(*) FROM turnos WHERE estado='CERRADO'");
    }

    public static List<Venta> ObtenerTodas()
    {
        using var conn = ConexionBD.Abrir();
        return conn.Query<Venta>("""
            SELECT v.id AS Id, v.turno_id AS TurnoId, v.cajero_id AS CajeroId,
                   u.nombre AS CajeroNombre, v.subtotal AS Subtotal,
                   v.iva AS Iva, v.total AS Total,
                   mp.nombre AS MetodoPagoNombre,
                   v.monto_recibido AS MontoRecibido, v.cambio AS Cambio,
                   v.estado AS Estado, v.creado_en AS CreadoEn
            FROM ventas v
            JOIN usuarios u ON u.id = v.cajero_id
            JOIN metodos_pago mp ON mp.id = v.metodo_pago_id
            ORDER BY v.creado_en DESC
            LIMIT 100
            """).ToList();
    }
}
